ops/testing.py

```diff
@@ -0,0 +1,372 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from textwrap import dedent
+
+from ops import charm, framework, model
+
+
+def create_harness(charm_cls, charm_meta_yaml):
+    """Used for testing your Charm or component implementations.
+
+    This ensures that you have an instance of `charm_cls` that can be driven by a TestingHarness.
+
+    Example::
+
+        harness = create_harness(MyCharm, '''
+            name: my-charm
+            requires:
+              db:
+                interface: pgsql
+            ''')
+        relation_id = harness.add_relation('db', 'postgresql')
+        harness.add_relation_unit(relation_id, 'postgresql/0', remote_unit_data={'key': 'value'})
+        # Check that charm has properly handled the relation_joined event for postgresql/0
+        self.assertEqual(harness.charm.
+
+    :param charm_cls: The Charm class that should be tested. If you are just testing a component,
+        you can pass in ops.charm.CharmBase.
+    :type charm_cls: CharmBase
+    :param charm_meta_yaml: The YAML metadata for the charm, defining interfaces, name, etc.
+        This can be either a string or a file.
+    :return: harness
+    :rtype: TestingHarness
+    """
+    # TODO: jam 2020-03-05 We probably want to take config as a parameter as well, since
+    #  it would define the default values of config that the charm would see.
+    if isinstance(charm_meta_yaml, str):
+        charm_meta_yaml = dedent(charm_meta_yaml)
+    meta = charm.CharmMeta.from_yaml(charm_meta_yaml)
+
+    # The Framework mutates class objects to build attributes for events, etc. That makes
+    # attribute access easy. However, it means you can't register the same class with
+    # multiple framework instances. So instead we dynamically create a new event class
+    # and charm class
+    # and register those with the framework.
+    class TestEvents(charm_cls.on.__class__):
+        pass
+
+    TestEvents.__name__ = charm_cls.on.__class__.__name__
+
+    class TestCharm(charm_cls):
+        on = TestEvents()
+
+    # Note: jam 2020-03-01 This is so that errors in testing say MyCharm has no attribute foo,
+    # rather than TestCharm has no attribute foo.
+    TestCharm.__name__ = charm_cls.__name__
+
+    unit_name = meta.name + '/0'
+    harness = TestingHarness(unit_name)
+    the_model = model.Model(unit_name, meta, harness._get_backend())
+    the_framework = framework.Framework(":memory:", "no-disk-path", meta, the_model)
+    the_charm = TestCharm(the_framework, meta.name)
+    # noinspection PyProtectedMember
+    harness._register_charm(the_charm)
+    return harness
+
+
+# noinspection PyProtectedMember
+class TestingHarness:
```

Comment:
This might be just Harness, so we have testing.Harness instead of testing.TestingHarness. Maybe we don't even need create_harness, actually, and instead just let people call it directly to instantiate it?